texts/0005-api-refactor.md

```diff
@@ -0,0 +1,79 @@
+
+# Guide-level explanation
+
+Two types of plugins will be defined: core plugins and regular (optional? external?) plugins. Core plugins are actually used by the `tauri` crate so they are defined in the crate itself and automatically configured and enabled, users will see these plugins as regular Rust modules that just export types and functions to be directly used. That is the case for the `path` and `file_system` plugins. The other APIs (`cli`,  `dialog`, `http`, `notification` and `process`) are optional and will be defined in the [plugins-workspace repo](https://github.com/tauri-apps/plugins-workspace) along with new plugins introduced in v2 like `camera` and `geolocation`.
+For the `path` plugin implementation, see [the PR](https://github.com/tauri-apps/tauri/pull/6339).
+  
+# Reference-level explanation
+
+Each API will follow the existing Tauri plugin structure, with the new mobile capabilities defined in the [RFC #0005](https://github.com/tauri-apps/rfcs/pull/9). Like any plugin in the plugins-workspace repository, they will have their own Rust crate and NPM package defining the JavaScript interface to consume its API, along with the Android and Swift packages implementing the functionality on mobile if needed.
+
+## Platform Considerations
+
+We can improve the existing desktop APIs, but it is not required to do so. The main goal of this refactor is to actually implement existing functionality on mobile, though improvements are always available and are being discussed for instance on the [capability-based API issue](https://github.com/tauri-apps/tauri/issues/6107).
+
+## Security Considerations
+
+With each API as a separate package, it will be easier to keep the security guarantees from previous audits. Changes will be easier to track and audit.
+
+## Performance Considerations
+
+Performance is not impacted by this change. Existing desktop code runs the same way, and mobile APIs are executed in the native layer so it can leverage the most performant way of doing things.
+
+## Community Considerations
+
+This refactor introduces a massive breaking change. All existing code using the `tauri::api` module or the `@tauri-apps/api` package would break. To mitigate this and simplify the migration path to v2, I propose we move all existing functionality (both the `tauri::api` module and the current Tauri endpoints) to a `legacy` plugin that can be enabled with a Cargo feature. With this plugin, an existing Tauri v1 application could start using the v2 functionalities by just enabling the Cargo feature and changing the `tauri::api` imports to `tauri::legacy`. This solution can also be used for other breaking changes for v2 and beyond, as existing functions that have incompatible new signatures could be moved to a Rust trait that can be imported by the developers to migrate major versions without touching their existing codebase.
```

Comment:
An alternative for this could be to create a migration script (we saw SvelteKit doing this for major breaking changes during beta). This would help to developer during the move from 1.x to 2.0. If there are other breaking changes they could possible be "fixed" with that script. 

I suggest this because I'm not sure if the proposed solution of using a cargo feature and a legacy module would be more to support from a maintainability perspective. I think having one path (i.e. only the new plugin API to import instead of both that and the legacy module) could be easier to communicate, explain, and troubleshoot.

Reply:
Capacitor also has some logic to migrate major versions. It might work on some simple cases like function renames, but for more complex breaking changes it might not be possible.